source/core_mqtt_serializer.c

```diff
@@ -1785,896 +3414,1985 @@ MQTTStatus_t MQTT_GetConnectPacketSize( const MQTTConnectInfo_t * pConnectInfo,
+                        status = decodeutf_8( &pServerRef, &pServerRefLength, &propertyLength,&serverRef, &pIndex );
+                        break;
+
+                    default:
+                        status = MQTTProtocolError;
+                        break;
+                }
+            }
+        }
+
+        return status;
     }
-    else
+
+/*-----------------------------------------------------------*/
+/*
+* API calls for Optional Subscribe Properties
+*/
+    MQTTStatus_t MQTTPropAdd_SubscribeId(MqttPropBuilder_t* pPropertyBuilder, size_t subscriptionId)
     {
-        switch( packetType )
+        MQTTStatus_t status = MQTTSuccess;
+        uint8_t* pIndex; 
+        if (subscriptionId == 0)
         {
-            /* Only publish acks are serialized by the client. */
-            case MQTT_PACKET_TYPE_PUBACK:
-            case MQTT_PACKET_TYPE_PUBREC:
-            case MQTT_PACKET_TYPE_PUBREL:
-            case MQTT_PACKET_TYPE_PUBCOMP:
-                pFixedBuffer->pBuffer[ 0 ] = packetType;
-                pFixedBuffer->pBuffer[ 1 ] = MQTT_PACKET_SIMPLE_ACK_REMAINING_LENGTH;
-                pFixedBuffer->pBuffer[ 2 ] = UINT16_HIGH_BYTE( packetId );
-                pFixedBuffer->pBuffer[ 3 ] = UINT16_LOW_BYTE( packetId );
-                break;
-
-            default:
-                LogError( ( "Packet type is not a publish ACK: Packet type=%02x",
-                            ( unsigned int ) packetType ) );
-                status = MQTTBadParameter;
-                break;
+            LogError(("Subscription Id cannot 0 for subscribe properties : Protocol Error "));
+            status = MQTTBadParameter;
+        }
+        else if ((pPropertyBuilder == NULL))
+        {
+            LogError(("Arguments cannot be NULL : pPropertyBuilder=%p.", (void*)pPropertyBuilder));
+            status = MQTTBadParameter;
+        }
+        else if (pPropertyBuilder->pBuffer == NULL)
+        {
+            LogError(("Arguments cannot be NULL : pPropertyBuilder->pBuffer=%p.", (void*)pPropertyBuilder->pBuffer));
+            status = MQTTBadParameter;
+        }
+        else if (UINT32_CHECK_BIT(pPropertyBuilder->fieldSet , MQTT_SUBSCRIPTION_ID_POS))
+        {
+            LogError(("Subscription Id already set"));
+            status = MQTTBadParameter;
+        }
+        else if (pPropertyBuilder->currentIndex + 1 + remainingLengthEncodedSize(subscriptionId) > pPropertyBuilder->bufferLength)
+        {
+            LogError(("Buffer too small to add subscription id"));
+            status = MQTTBadParameter;
+        }
+        else {
+            pIndex = pPropertyBuilder->pBuffer + pPropertyBuilder->currentIndex;
+            *pIndex = MQTT_SUBSCRIPTION_ID_ID;
+            pIndex++;
+            pIndex = encodeRemainingLength(pIndex, subscriptionId);
+            pPropertyBuilder->currentIndex += (size_t)(pIndex - (pPropertyBuilder->pBuffer + pPropertyBuilder->currentIndex));
```

Comment:
We need checks to make sure that we are not going out of bound of the `pPropertyBuilder->pBuffer` by verifying `pPropertyBuilder->currentIndex` and the required encoding length with `pPropertyBuilder->bufferLength`.
This can be added as a separate preliminary check for all similar APIs with a separate error code.